libmamba/src/core/download.cpp

```diff
@@ -0,0 +1,609 @@
+                build_transfer_message(res.transfer.http_status, res.transfer.effective_url, res.transfer.downloaded_size));
+        }
+    }
+
+    void DownloadTracker::throw_if_required(const DownloadError& res)
+    {
+        if (m_state == DownloadState::FAILED && !p_request->ignore_failure)
+        {
+            throw std::runtime_error(res.message);
+        }
+    }
+    
+    /*****************************
+     * DOWNLOADER IMPLEMENTATION *
+     *****************************/
+
+    Downloader::Downloader(MultiDownloadRequest requests, DownloadOptions options, const Context& context)
+        : m_requests(std::move(requests))
+        , m_options(std::move(options))
+        , p_context(&context)
+        , m_curl_handle(context.threads_params.download_threads)
+        , m_trackers()
+    {
+        if (m_options.sort)
+        {
+            std::sort(
+                m_requests.requests.begin(),
+                m_requests.requests.end(),
+                [](DownloadRequest& a, DownloadRequest& b) -> bool
+                { return a.expected_size.value_or(SIZE_MAX) > b.expected_size.value_or(SIZE_MAX); }
+            );
+        }
+
+        m_trackers.reserve(m_requests.requests.size());
+        std::size_t max_retries = static_cast<std::size_t>(context.remote_fetch_params.max_retries);
+        bool fail_fast = options.fail_fast;
+        std::transform(m_requests.requests.begin(), m_requests.requests.end(), std::inserter(m_trackers, m_trackers.begin()),
+            [max_retries, fail_fast](const DownloadRequest& req) { return DownloadTracker(&req, max_retries, fail_fast); });
+        m_waiting_count = m_trackers.size();
+    }
+    
+    MultiDownloadResult Downloader::download()
+    {
+        while(!download_done())
+        {
+            prepare_next_downloads();
+            update_downloads();
+        }
+
+        return build_result();
+    }
+
+    void Downloader::prepare_next_downloads()
+    {
+        size_t running_attempts = m_completion_map.size();
+        size_t max_parallel_downloads = p_context->threads_params.download_threads;
+        auto start_filter = mamba::util::filter(m_trackers, [&](DownloadTracker& tracker)
+        {
+            return running_attempts < max_parallel_downloads && tracker.can_start_transfer();
+        });
+
+        for (auto& tracker: start_filter)
+        {
+            auto [iter, success] = m_completion_map.insert(tracker.prepare_new_attempt(m_curl_handle, *p_context));
+            if (success)
+            {
+                ++running_attempts;
+            }
+        }
+    }
+
+    void Downloader::update_downloads()
+    {
+        while(auto resp = m_curl_handle.pop_message())
+        {
+            const auto& msg = resp.value();
+            if (!msg.m_transfer_done)
+            {
+                // We are only interested in messages about finished transfers
+                continue;
```

Comment:
does `pop_message()` handles the waiting of the current thread? I would use `std::yield()` here otherwise, to avoid saturating the cpu.

Reply:
nope this should be done in `m_curl_handle->perform()` actually, which is missing from this function.